resource.go

```diff
@@ -25,24 +25,46 @@ func resourceNameFromStruct(v interface{}) string {
 	return strings.ToLower(structs.Name(v))
 }
 
+func resourceNameFromMap(v interface{}) (string, error) {
+	if m, ok := v.(map[string]interface{}); !ok {
+		return "", errors.New("Invalid map type")
+	} else if val, ok := m["pinejs"].(string); ok && val != "" {
+		return val, nil
+	} else {
+		return "", errors.New("Failed to retrieve resource name from map")
+	}
+}
+
 // Unwinds pointers, slices, and slices of pointers, etc. until we get to a
-// struct then we hand off to resourceNameFromStruct.
+// struct then we hand off to resourceNameFromStruct, or the equivalent with a map.
 func resourceName(v interface{}) (string, error) {
 	ty := reflect.TypeOf(v)
 
 	switch ty.Kind() {
 	case reflect.Struct:
 		return resourceNameFromStruct(v), nil
-	case reflect.Ptr, reflect.Slice:
-		// Create new pointer to pointer/slice type.
-		ptr := reflect.New(ty.Elem())
-
-		// Deref the pointer and recurse on that value until we get to a struct.
-		el := ptr.Elem().Interface()
-		return resourceName(el)
+	case reflect.Map:
+		return resourceNameFromMap(v)
+	case reflect.Ptr:
+		return resourceName(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	case reflect.Slice:
+		if ty.Elem().Kind() == reflect.Struct {
```

Comment:
It'd be nice to separate this logic out into a function as with the other cases + make this a little more readable.